crates/roc_host/src/json_web_token.rs

```diff
@@ -0,0 +1,281 @@
+use roc_std::{RocList, RocStr};
+use std::collections::HashMap;
+
+#[repr(C)]
+pub struct FromRocJwt {
+    secret: RocStr,
+    token: RocStr,
+}
+
+impl roc_std::RocRefcounted for FromRocJwt {
+    fn inc(&mut self) {
+        self.secret.inc();
+        self.token.inc();
+    }
+    fn dec(&mut self) {
+        self.secret.dec();
+        self.token.dec();
+    }
+    fn is_refcounted() -> bool {
+        true
+    }
+}
+
+#[repr(C)]
+pub struct ToRocJwtClaims {
+    name: RocStr,
+    value: RocStr,
+}
+
+impl roc_std::RocRefcounted for ToRocJwtClaims {
+    fn inc(&mut self) {
+        self.name.inc();
+        self.value.inc();
+    }
+    fn dec(&mut self) {
+        self.name.dec();
+        self.value.dec();
+    }
+    fn is_refcounted() -> bool {
+        true
+    }
+}
+
+/// Represent the claims as an equivalent RocStr json value
+fn jwt_claims_to_roc(claims: HashMap<String, serde_json::Value>) -> RocList<ToRocJwtClaims> {
+    let mut roc_claims = RocList::with_capacity(claims.len());
+    let json_to_str = |json_value: serde_json::Value| -> RocStr {
+        match json_value {
+            serde_json::Value::String(s) => s.as_str().into(),
+            serde_json::Value::Number(n) => format!("{}", n).as_str().into(),
+            serde_json::Value::Bool(b) => {
+                if b {
+                    "true".into()
+                } else {
+                    "false".into()
+                }
+            }
+            serde_json::Value::Null => "null".into(),
+            _ => panic!("unsupported json type"),
+        }
+    };
+
+    for (name, json_value) in dbg!(claims) {
+        roc_claims.push(ToRocJwtClaims {
+            name: name.as_str().into(),
+            value: json_to_str(json_value),
+        });
+    }
+
+    roc_claims
+}
+
+fn jwt_error_to_roc(err: jwt::error::Error) -> ToRocJwtErr {
+    match err {
+        jwt::error::Error::AlgorithmMismatch(..) => ToRocJwtErr::algorithm_mismatch(),
+        jwt::error::Error::InvalidSignature => ToRocJwtErr::invalid_signature(),
+        jwt::error::Error::Format => ToRocJwtErr::bad_format(),
+        jwt::error::Error::Json(..) => ToRocJwtErr::bad_json(),
+        jwt::error::Error::NoClaimsComponent => ToRocJwtErr::missing_claims(),
+        jwt::error::Error::NoHeaderComponent => ToRocJwtErr::missing_header(),
+        jwt::error::Error::NoSignatureComponent => ToRocJwtErr::missing_signature(),
+        jwt::error::Error::NoKeyId => ToRocJwtErr::missing_key_id(),
+        _ => ToRocJwtErr::other(err.to_string().as_str().into()),
+    }
+}
+
+pub fn jwt_verify(jwt: &FromRocJwt) -> Result<RocList<ToRocJwtClaims>, ToRocJwtErr> {
+    let header = json_web_token::jwt_header(&jwt.token).map_err(|_| ToRocJwtErr::bad_format())?;
+
+    match header.algorithm {
+        jwt::AlgorithmType::Hs256 => json_web_token::jwt_verify_hs256(&jwt.token, &jwt.secret)
+            .map(jwt_claims_to_roc)
+            .map_err(jwt_error_to_roc),
```

Comment:
You technically could move these two lines outside of the match. Then you don't have to repeat them in every branch.